File: code/graph_algorithms/src/floyd_warshall_algorithm/floyd_warshall_algorithm.py
""" Part of Cosmos by OpenGenus Foundation """

INF = 1000000000


def floyd_warshall(vertex, adjacency_matrix):

    # calculating all pair shortest path
    for k in range(0, vertex):
        for i in range(0, vertex):
            for j in range(0, vertex):

                # relax the distance from i to j by allowing vertex k as intermediate vertex
                # consider which one is better, going through vertex k or the previous value
                adjacency_matrix[i][j] = min(
                    adjacency_matrix[i][j],
                    adjacency_matrix[i][k] + adjacency_matrix[k][j],
                )

                # pretty print the graph
                # o/d means the leftmost row is the origin vertex
                # and the topmost column as destination vertex
    print("o/d", end="")
    for i in range(0, vertex):
        print("\t{:d}".format(i + 1), end="")
    print()

    for i in range(0, vertex):
        print("{:d}".format(i + 1), end="")

        for j in range(0, vertex):
            print("\t{:d}".format(adjacency_matrix[i][j]), end="")

        print()


"""
input is given as adjacency matrix,
input represents this undirected graph

 A--1--B
 |    /
 3   /
 |  1
 | /
 C--2--D

should set infinite value for each pair of vertex that has no edge
 """

adjacency_matrix = [[0, 1, 3, INF], [1, 0, 1, INF], [3, 1, 0, 2], [INF, INF, 2, 0]]
floyd_warshall(4, adjacency_matrix)

File: code/graph_algorithms/src/floyd_warshall_algorithm/floyd_warshall_algorithm.cs
using System;

namespace Cosmos
{
	public static class FloydWarshall
	{
		private const float Infinity = float.PositiveInfinity;

		public static void Solve(float[,] graph)
		{
			var v = graph.GetLength(0);

			for (var k = 0; k < v; k++)
			{
				for (var i = 0; i < v; i++)
				{
					for (var j = 0; j < v; j++)
					{
						graph[i, j] = Math.Min(graph[i, j], graph[i, k] + graph[k, j]);
					}
				}
			}

			for (int i = 0; i < v; i++)
			{
				for (int j = 0; j < v; j++)
				{
					Console.Write(float.IsPositiveInfinity(graph[i, j]) ? "∞ " : $"{graph[i, j]} ");
				}
				Console.WriteLine();
			}
		}
		
		private static void Main(string[] args)
		{
			var graph = new float[,]
			{
				{0, 5, Infinity, 10},
				{Infinity, 0, 3, Infinity,},
				{Infinity, Infinity, 0, 1},
				{Infinity, Infinity, Infinity, 0}
			};
			
			FloydWarshall.Solve(graph);
		}
	}
}